### PR TITLE
CMake: Add improvements to MPI support

### PR DESCRIPTION
**Pull request by johnwparent**
This PR vendors a few related improvements to MPI support in the CMake buildsystem on the Windows platform.

(1) Plugins may be dependent on MPI includes, ensure these explicitly happen in the CMake build system. Without this, compilation fails in the HDF5 plugins with HDF5 enabled and built with MPI on Windows.

(2) When linking, certain components of the final `netcdf::netcdf` target are reliant on the inclusion of MPI libraries in the linker line, establish this directly with a CMake target link interface addition for `MPI::MPI_C`. If this is not done, linking fails on Windows platforms building with parallel HDF5 and parallelism enabled.

(3) When exporting the project, if a third party dependency (here MPI) is going to be included in a link interface, it must also be found by the exporting project. Since (2) adds MPI to the link interface, we must then find MPI in the exported netCDFConfig.cmake. Without this, and with (2), consuming projects will fail to generate as the MPI target denoted in the link interface for `netcdf::netcdf` will not be present in the consuming project unless that project also brings in MPI.

Note: These changes have only been tested on the Windows platform.

### CODE REVIEW

**Comment by WardF** [comment]
Thanks @johnwparent, I'll take a look at this! Would you mind accepting the CLA as well? Thanks!

File: liblib/CMakeLists.txt [hunk 1]
```diff
@@ -52,6 +52,10 @@ IF(MPI_C_INCLUDE_PATH)
     target_include_directories(netcdf PUBLIC ${MPI_C_INCLUDE_PATH})
 ENDIF(MPI_C_INCLUDE_PATH)
 
+IF(TARGET MPI::MPI_C)
+  target_link_libraries(netcdf MPI::MPI_C)
+ENDIF(TARGET MPI::MPI_C)
```
Comment (ZedThree) [inline liblib/CMakeLists.txt:57]: Linking against the target like this means all the include directories are also added -- which means L51-53 can go.

I'd probably also suggest making this link public:

```suggestion
IF(TARGET MPI::MPI_C)
  target_link_libraries(netcdf PUBLIC MPI::MPI_C)
ENDIF(TARGET MPI::MPI_C)
```

File: plugins/CMakeLists.txt [hunk 2]
```diff
@@ -62,6 +62,9 @@ MACRO(buildplugin TARGET TARGETLIB)
     set_target_properties(${TARGET} PROPERTIES LINK_FLAGS "/INCREMENTAL:NO /DEBUG /OPT:REF /OPT:ICF")
     # Set file name & location
     set_target_properties(${TARGET} PROPERTIES COMPILE_PDB_NAME ${TARGET} COMPILE_PDB_OUTPUT_DIR ${CMAKE_BINARY_DIR})
+    IF(MPI_C_INCLUDE_PATH)
+      target_include_directories(${TARGET} PRIVATE ${MPI_C_INCLUDE_PATH})
+    ENDIF(MPI_C_INCLUDE_PATH)
```
Comment (ZedThree) [inline plugins/CMakeLists.txt:67]: Better to use `target_link_libraries` with the MPI target to make sure this gets linked against the libraries if necessary:

```suggestion
    IF(TARGET MPI::MPI_C)
      target_link_libraries(${TARGET} PRIVATE MPI::MPI_C)
    ENDIF(TARGET MPI::MPI_C)
```

**Comment by ZedThree** [comment]
I found some of these changes necessary to compile netCDF successfully on Linux too. I found changing all the `target_include_directories/target_link_libraries` to

a) use the `MPI::MPI_C` target
b) change the conditional to `STATUS_PARALLEL`

to work best on my system, for example:

```diff
-IF(MPI_C_INCLUDE_PATH)
-    target_include_directories(nczarr PUBLIC ${MPI_C_INCLUDE_PATH})
-ENDIF(MPI_C_INCLUDE_PATH)
+IF(STATUS_PARALLEL)
+    target_link_libraries(nczarr PUBLIC MPI::MPI_C)
+ENDIF(STATUS_PARALLEL)
```

File: liblib/CMakeLists.txt (hunk 1 above)
Comment (johnwparent) [reply liblib/CMakeLists.txt:57]: Specifying a public link here would require a refactor of more of the CMake codebase as mixing signatures to `target_link_libraries` for a given target is not permitted by CMake.

The netcdf target is linked later in the list file via the "plain-signature" so it cannot be linked with the keyword signature. If there is support for refactoring all of the incidences of the `netcdf` targets link interface, then I'd be happy to do so.

Comment (ZedThree) [reply liblib/CMakeLists.txt:57]: Ah yes, good point, forgot about that!

**Comment by WardF** [comment]
@dopplershift

**Comment by dopplershift** [comment]
Here's the ["entity" CLA](https://gist.github.com/dopplershift/957a0b9d1c713ad1ad765c8e4a79e3c1).

**Comment by johnwparent** [comment]
@dopplershift Thanks! I've forwarded that along. How should I return the signed contract to you folks?

**Comment by dopplershift** [comment]
@johnwparent A PDF would be fine.

**Comment by WardF** [comment]
Getting caught up on the backlog; did this get resolved, @dopplershift ?

**Comment by johnwparent** [comment]
@WardF I am unfortunately still waiting on one of the stakeholders to provide signoff, and unfortunately they move a bit slowly in the legal space. I will update here when I am able to provide the signed entity CLA.

**Comment by WardF** [comment]
Just checking in here again re CLA

**Comment by johnwparent** [comment]
Thanks, I just reached back out to my contracting contact point, hopefully that will inspire some movement.

**Comment by mathstuf** [comment]
Related to #2742 where configure checks in CMake fail when HDF5 is built with MPI.

**Comment by mathstuf** [comment]
FWIW, I think that instead of testing for variable existence, instead check `HDF5_PARALLEL` which is what we're really interested in here (e.g., if these path variables get set as a cache variable, it stays that way even if MPI support is later turned off). Ideally, HDF5's exported targets would just bring along the MPI usage requirements, but I suspect that's not going to be fixed anytime soon across all the ways HDF5 supports being built/installed.

**Pull request merged**